app/methods/tag_adder.rb

```diff
@@ -1,24 +1,39 @@
 class TagAdder
-  def initialize(prediction:, string:)
+  def initialize(prediction:, save: true, string:)
     @prediction = prediction
-    @string = string
+    @save = save
+    @string = string || ''
   end
 
   def call
     if string.present? && prediction.present?
       tag_names.each do |tag_name|
 	      prediction.tag_names << tag_name
-      end && prediction.save
+      end && save && prediction.save
```

Comment:
I find these "buried side effects" really cringeworthy now I work in Haskell most of the time. Would it be too much to ask that this class just modifies the prediction without persisting it, and it's up to the caller to decide whether or not to persist the modifications?

Reply:
Agreed. Let me rework this.

Reply:
[This](https://github.com/bellroy/predictionbook/pull/206/commits/9c1b284ae9d69c7d2b1af53138fce6ffb15299fa) commit gets `TagAdder` out of the prediction persistence business and also eliminates the need for the Boolean parameter.